### PR TITLE
Small Cata fixes

### PR DESCRIPTION
**Pull request by TyThompson**
It isn't perfect but it is a step in the right direction

Last bug I'm looking at is
AutoGear/AutoGear.lua:2101: attempt to index local 'realSpec' (a number value)
[string "@AutoGear/AutoGear.lua"]:2101: in function <AutoGear/AutoGear.lua:2079>

### CODE REVIEW

**Review by AlexFolland — CHANGES_REQUESTED** [review]
I just have 1 comment which requests a (potentially) minor change, but otherwise, this looks good to me.  Thanks!

**Comment by AlexFolland** [comment]
Also, I'm wondering why it's necessary to introduce an anonymous function that just calls `AutoGearMain` instead of just passing `AutoGearMain` to `AutoGearFrame:SetScript` directly like we've always done before this.  What changed or what did you observe to make that necessary?

I realize the difference is literally just an additional function call on every update and this almost doesn't matter, but I am curious.

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:2206]: Disenchant roll being temporarily not in the game source: https://youtube.com/watch?v=f8zWAPDUTkc&t=2498s

**Comment by TyThompson** [comment]
It looked strange to me so I made it a one liner like the lines above it. The old code works as well.

**Comment by AlexFolland** [comment]
> It looked strange to me so I made it a one liner like the lines above it. The old code works as well.

OK, please change that part back then.  We don't need to change that part.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:2101]: I looked at what's happening here, and it looks like Cataclysm has a different first return value for `GetTalentTabInfo`.  So, the `AutoGearDetectSpec` implementation for Cataclysm needs to select the second return value instead of the first for the spec name instead of calling `tostring` here, I think.  What do you think?

I say this without being able to test this (as I'm still not at my home PC with WoW available) because I did a github search for GetTalentTabInfo and found this line in the BigWigs code:

https://github.com/BigWigsMods/BigWigs/blob/53e8515d7c043e48714798cb16d9643eb818ab02/Core/BossPrototype_Classic.lua#L105

This indicates that BigWigs checks for Cataclysm Classic and selects a different return value because of that.  So, AutoGear should do the same.  Looks like the spec name string is the second return value instead of the first in Cataclysm Classic.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Even though I can't test because I'm still not at my home PC, I think I found the root cause of spec being a number instead of a string.  I've commented on that.  Thanks again for your latest changes.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:1613]: I think this line needs ` or nil` appended for the case where `AutoGearDefaultWeights[class]` is not `nil` but `AutoGearDefaultWeights[class][spec]` is `nil`.

By the way, what does this change accomplish?  I thought the original line would result in `nil` even in the case where `AutoGearDefaultWeights[class]` is `nil`.  Was I incorrect there?

Also, could this whole new block be written in a one-liner like this instead, like the `weapons` assignment below it?

```Lua
AutoGearCurrentWeighting = AutoGearDefaultWeights[class] and (AutoGearDefaultWeights[class][spec] or nil) or nil
```

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Thanks for more changes.  I have another review for you!  This is all looking like great progress.  Sorry for being unable to test right now.  I should have access to my home PC some time later this week or next week.

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:4313]: Still working on this one

It isn't showing the Tooltip but it isn't breaking the client either. You are able to still use the addon in Dragonflight and SoD/Cata

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:4302]: I believe this would be Interface: 100002 to be precise
https://wowpedia.fandom.com/wiki/Patch_10.0.2/API_changes#Tooltip_Changes

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:2147]: I'm ok with that I was struggling on this one

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:4309]: That looks very familiar.  I think that exact code is part of what I had tried when I last worked on this, but it didn't work.

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:2173]: recent commit fixes the underlying problem

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua]: -- changed in Dragonflight 10.2.6
-- source: https://warcraft.wiki.gg/wiki/API_C_Item.GetItemInfo
-- source: https://warcraft.wiki.gg/wiki/API_C_Item.IsEquippedItem

Changes for GetItemInfo and IsEquippedItem as well

**Review by wfwfewe — APPROVED** [review]

**Comment by AlexFolland** [comment]
Why was this closed?  Who is @wfwfewe?  I've just now changed the settings to prevent non-collaborators from reviewing.  I'll be checking this out this weekend.

**Comment by TyThompson** [comment]
Sorry first pull request I was confused after it was approved lol

**Comment by AlexFolland** [comment]
Oh, that makes sense!  Haha, yeah, I was confused when a non-contributor with no contributions anywhere on the site was allowed to submit a code review on this project.  I became aware of a setting that prevents that as a result.

Sorry, the reality of family life has meant that I haven't had any time this weekend to boot up WoW and play with this change.  I plan to do it soon still, but having 2 kids makes free time very limited.  I can come on here and comment once in a while when I get a couple minutes to sit down, but dedicating actual time is just not realistic sometimes.

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua]: Since this change does not resolve the tooltip issue, I think these Dragonflight-specific fixes should be removed from this pull request and separated into a different pull request.  That would require having the changes in a different branch in your fork.

File: AutoGear-Mainline.toc
Comment (AlexFolland) [inline AutoGear-Mainline.toc:1]: I think this should be moved to a separate branch, to be added to a different mainline-specific pull request.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:73]: I would like to avoid the maintenance burden this system implies.  If we use any of these booleans, behavior changes based on TOC instead of function interfaces.  In the case you want to use `isCata` here for, we should do something more dynamic like checking the number of return values or the return type of `GetTalentTabInfo` instead.  That way, it won't stop working in the future, like a "Classic MoP", for example.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:276]: I would like to avoid the maintenance burden this implies.  If we check `isCata` here, behavior changes based on TOC instead of function interfaces.  In this case, I think we should do something more dynamic like checking the number of return values or the return type of `GetTalentTabInfo` instead.  That way, it won't stop working in the future, like a "Classic MoP", for example.

Maybe something like this:

```Lua
local _, spec, _, _, pointsSpent = GetTalentTabInfo(1)
if pointsSpent and pointsSpent >= 0 then
	highestPointsSpent = pointsSpent
	for i = 2, numTalentTabs do
		local _, spec, _, _, pointsSpent = GetTalentTabInfo(i)
		if (pointsSpent > highestPointsSpent) then
			highestPointsSpent = pointsSpent
			highestSpec = spec
		end
	end
else
	for i = 1, numTalentTabs do
		local spec, _, pointsSpent = GetTalentTabInfo(i)
		if (highestPointsSpent == nil or pointsSpent > highestPointsSpent) then
			highestPointsSpent = pointsSpent
			highestSpec = spec
		end
	end
end
```

With this, we'd try the 5-return call and if it's not nil and `pointsSpent` is a positive number, then it's pretty safe to say that it's using that 5-return-value implementation regardless of what expansion it is.  Otherwise, it'll use the 3-return-value implementation which still works elsewhere.

File: AutoGear.lua
Comment (AlexFolland) [reply AutoGear.lua:2174]: I see this assignment is still here, and I don't think it's necessary since `realSpec` is already a string.

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:4334]: I think the Dragonflight-specific fixes should be removed from this pull request and separated into a different pull request.  That would require having the changes in a different branch in your fork.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Here's my review of the latest changes.  I still haven't tested in-game, but my main points are that:

- I think the Dragonflight-specific changes should be in another pull request.
  - My rationale: The Dragonflight-specific changes complicate this pull request as they're not complete and completing them has different satisfaction requirements than completing the Cataclysm-specific changes.
  - I believe the Cataclysm-specific changes are almost ready, and separating the Dragonflight-specific changes out would allow this to finish earlier.
- Even though I like how you've written them if they were needed, I would prefer to avoid adding and being tempted to use a set of expansion-specific booleans like `isCata`.
  - Rationale: Interpreting the number of return values dynamically is likely to be more future-proof and less of a maintenance burden than just checking a particular expansion as enablement for a specific code path.

File: AutoGear-Mainline.toc
Comment (TyThompson) [reply AutoGear-Mainline.toc:1]: going back to 100000

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:73]: dropping the isExpansion

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua:4334]: drop the tostring

**Comment by TyThompson** [comment]
Made the requested changes

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua]: implemented your code

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua]: took the dragonflight stuff out

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua]: dropped realspec to string

File: AutoGear.lua
Comment (AlexFolland) [inline AutoGear.lua:255]: Reading this again, I realized there's a bug in my suggested code which you faithfully copied.  After the first `highestPointsSpent = pointsSpent`, it doesn't run `if pointsSpent > 0 then highestSpec = spec end`.  I think that line needs to be added, just to make sure that if there's at least 1 point spent in talent tree 1, that tree is marked as the spec tree.

**Review by AlexFolland — CHANGES_REQUESTED** [review]
Just 1 more small bug to fix which was in my suggested code, and then this looks good to me!

File: AutoGear.lua
Comment (TyThompson) [reply AutoGear.lua]: added highestSpec = spec

Comment (AlexFolland) [reply AutoGear.lua]: It needs a condition of being above 0 or else it will assign highestSpec to the first talent tree even if there are 0 points in it.

Comment (TyThompson) [reply AutoGear.lua]: Updated

Comment (AlexFolland) [reply AutoGear.lua]: This latest change isn't right either, because the check there can't be combined.  Checking for a number including 0 is to check for whether the value is a number at all and enters the correct block for the number of GetTalentTabInfo return values.

**Comment by AlexFolland** [comment]
I'm merging this and fixing the last issue myself.  Thank you very much for your effort!

**Pull request merged**
